test: add tests for CommentInput

dyesseyumba opened the pull request:
Hi, here is the unit test for the component 'CommentInput'.

alejandronanez commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,132 @@
+
+    wrapper.instance().forceUpdate();
+
+    Platform.OS = 'android';
```
You should mock this instead of modifying it directly.

dyesseyumba replied:
I failed to import the mocked `Platform` object in wrapped `CommentInput`. 

CommentInput import the `Platform` from `'react-native'`. The challenge here is to inject the mocked Platform to CommentInput when called in import. But I didn't succeed. For that, I think I should use [inject-loader](https://www.jamestease.co.uk/blether/mocking-es6-imports-for-tests).

Do you have any idea?

machour commented:
@dyesseyumba I'm curious about how you merge HEAD into your branch.. it seems to generate a lot of noise every time (look at the number of commits in this PR vs the last one on #533 when I merged HEAD).
I would happily help you figure that out if you pop by on gitter!

dyesseyumba commented:
@machour You're right. There must be something I'm doing wrong. I'm postin on gitter.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+    expect(wrapper.find('Text').length).toEqual(0);
+  });
+
+  it("should not render 'TextInput' and TouchableOpacity if user can post", () => {
```
if user can't post

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+      <CommentInput
+        {...defaultProps}
+        userHasPushPermission={false}
+        issueLocked={true}/>
```
Shall we separate as 2 cases? For `userHasPushPermission` and `issueLocked`.

dyesseyumba replied:
I think no. Because `userHasPushPermission ` and `issueLocked` are linked with the operation and.
![image](https://user-images.githubusercontent.com/5528931/32133075-8c3cc0a4-bbc8-11e7-91a3-5a5995dc138a.png)

chinesedfan replied:
@dyesseyumba Yes, there are 2 kinds of results. But we have 4 kinds of inputs, `userHasPushPermission` x2 and `issueLocked` x2. And `issueLocked` also affects the locked issue text and icon. Can you cover them?

dyesseyumba replied:
Okay, I'm on it.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+
+describe('<CommentInput />', () => {
+
+  let defaultProps = {
```
const

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+      <CommentInput
+        {...defaultProps}
+        userHasPushPermission={true}
+        issueLocked={false}/>
```
Are these 2 props the same with `defaultProps`?

dyesseyumba replied:
Yes, there are. I'll use `defaultProps`.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+      <CommentInput
+        {...defaultProps}
+        userHasPushPermission={true}
+        issueLocked={false}/>
```
Same here.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+      <CommentInput
+        {...defaultProps}
+        userHasPushPermission={true}
+        issueLocked={false}/>
```
Same here.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+      <CommentInput
+        {...defaultProps}
+        userHasPushPermission={true}
+        issueLocked={false}/>
```
Same here.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+
+    expect(handleSubmitEditingSpy).toHaveBeenCalled();
+
+    expect(wrapper.state('text')).toEqual('Changed by submitEditing\n');
```
Sorry, why does it end with `\n`?

dyesseyumba replied:
It is because the character `\n` is used in the method handleSubmitEditing.
![image](https://user-images.githubusercontent.com/5528931/32132895-510aa0e0-bbc4-11e7-8773-fec8e70596d3.png)

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+      .instance()
+      .forceUpdate();
+
+    Platform.OS = 'android';
```
Can you explain this line for me? I am curious about `iOS`.

dyesseyumba replied:
There is an instruction for the platform Android in the method `handleSubmitEditing`. And the `Platform` object in imported in the class `CommentInput` from `'react-native'`. But I failed to mock and import the platform object in `CommentInput`. I can use `inject-loader` as explained in [this article](https://www.jamestease.co.uk/blether/mocking-es6-imports-for-tests) but I have to install it.
![image](https://user-images.githubusercontent.com/5528931/32132895-510aa0e0-bbc4-11e7-8773-fec8e70596d3.png)

About `iOS`; there is no condition for that in  `CommentInput` class

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,131 @@
+    onSubmit: () => {},
+  };
+
+  it("should render 'TextInput' and TouchableOpacity if user can post", () => {
```
Maybe no single quotes or both use for `TextInput` and `TouchableOpacity`.

andrewda commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,119 @@
+    onSubmit: () => {},
+  };
+
+  it('should render TextInput and TouchableOpacity if I can post', () => {
```
Use 3rd person. `'should render TextInput and TouchableOpacity if user has post permissions'`

andrewda commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,119 @@
+    expect(wrapper.find('Text').length).toEqual(0);
+  });
+
+  it("should not render TextInput and TouchableOpacity if I can't post", () => {
```
Use 3rd person. `'should not render TextInput and TouchableOpacity if user does not have post permissions'`

andrewda commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,119 @@
+    expect(wrapper.find('Text').length).toEqual(1);
+  });
+
+  it('should update the state text if value changed', () => {
```
`'should update the state text if value is changed'`

andrewda commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,119 @@
+    expect(wrapper.state('text')).toEqual('Changed text');
+  });
+
+  it('should call handleSubmit methods when submitted', () => {
```
`'should call handleSubmit method when submitted'`

andrewda commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,119 @@
+    expect(wrapper.state('height')).toBe(10);
+  });
+
+  it('should call handleSubmitEditing methods when onSubmitEditing event raised', () => {
```
`'should call handleSubmitEditing method when onSubmitEditing event is raised'`

chinesedfan replied:
@dyesseyumba `'should call handleSubmitEditing method in Android when onSubmitEditing event is raised'`. Please also mention the platform.

chinesedfan commented:
@dyesseyumba Thanks for your updates. Please notice that more components were introduced, but you didn't update test descriptions. It is better to only test one component or some related components in each test case.

dyesseyumba commented:
@chinesedfan I'm a little lost about new components. Those components are not the olds components styled with `'styled-components'`?  I updated the test description by adding the prefix "styled" in front of the component name. For example for the component `InputText` I used `styled TextInput` in description because it appear as `Styled(TextInput)` with enzyme.
Do I have to use the name of the component as it appears in ComponentInput class?

You right about to only test one component. I added unnecessary assertions for child components. I'll fix that.

chinesedfan commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,151 @@
+    expect(wrapper.find('Styled(TextInput)').length).toEqual(1);
+    expect(wrapper.find('Styled(TouchableOpacity)').length).toEqual(1);
+    expect(wrapper.find('Styled(PostButtonIcon)').length).toEqual(1);
+    expect(wrapper.find('Styled(Text)').length).toEqual(0);
```
The description didn't mention  "Styled(PostButtonIcon)" and "Styled(Text)". They are what I mean "more components".

dyesseyumba commented on __tests__/tests/components/CommentInput.js:
```diff
@@ -0,0 +1,184 @@
+    expect(wrapper.find('Styled(TouchableOpacity)').length).toEqual(1);
+  });
+
+  it('should not render styled Text and icon if user does not have post permissions and issue is not locked', () => {
```
Sorry! You're right, I missed that. I'll take my time.

dyesseyumba commented:
@andrewda and @alejandronanez I applied all changes you requested but you didn't approve. Is anything wrong with this PR?

chinesedfan commented:
Ping @andrewda and @alejandronanez. Seems good to me now.

chinesedfan commented:
> Don't hardcode Platform

Except the above suggestion by @alejandronanez, everything is OK. But I think it is still acceptable and this PR has been pending for such long time without new feedbacks. So let it be merged!
